messaging/src/main/protos/Messages.proto

```diff
@@ -0,0 +1,185 @@
+ * an authentication challenge to the client and the client responds with a signed authentication
+ * response identifying its identityKey and deviceId. On anonymous connections, clients simply ignore the
+ * authentication challenge.
+ *
+ * Messages sent from clients to servers follow a request/response pattern. The server will always
+ * respond to these with either an Ack or a typed response. In the event of an error, it will respond
+ * with an Error message. This includes the following messages:
+ *
+ *  - Register        -> Ack
+ *  - Unregister      -> Ack
+ *  - RequestPreKeys  -> PreKeys
+ *  - OutboundMessage -> Ack
+ *
+ * Some messages sent from the server to the client require an Ack in response:
+ *
+ *  - inboundMessage  -> Ack
+ *
+ * Some messages don't require any response:
+ *
+ *  - PreKeysLow
+ *
+ * All messages sent within a given connection are identified by a unique sequence number (separate sequences
+ * for each direction). When a response message is sent in either direction, its sequence number is set to the
+ * message that triggered the response so that the client or server can correlate responses with requests.
+ */
+syntax = "proto3";
+
+package tassis;
+
+option go_package = "github.com/getlantern/tassis/model";
+option java_package = "io.lantern.messaging.tassis";
+
+// Acknowledges successful receipt of a Message
+message Ack {}
+
+// Indicates than an error occurred processing a request.
+message Error {
+  string name        = 1;  // An identifier for the error, like "unknown_identity"
+  string description = 2;  // Optional additional information about the error
+}
+
+// An Address for a specific client
+message Address {
+  bytes identityKey = 1;  // The 32 byte ed25519 public key that uniquely identifies an identity (e.g. a user)
+  bytes deviceId    = 2;  // Identifier for a specific device, only unique for a given identityKey
+}
+
+// Provides configuration information to clients
+message Configuration {
+  int64 maxAttachmentSize = 1; // The maxmimum allows attachment size (encrypted size, not plaintext)
+}
+
+// Login information supplied by clients in response to an AuthChallenge.
+message Login {
+  Address address = 1;  // The Address that's logging in. This will become permanently associated with the current connection
+  bytes   nonce   = 2;  // This echoes back the nonce provided by the server in the AuthChallenge
+}
+
+// A challenge to the client to authenticate. This is sent by the server once and only once, immediately after clients connect.
+message AuthChallenge {
+  bytes nonce = 1;  // A nonce to identify this authentication exchange
+}
+
+// A response to an AuthChallenge that is sent from the client to the server on any connection that the client wishes to authenticate.
+// The server will accept an AuthResponse only once per connection.
+message AuthResponse {
+  bytes login     = 1;  // The serialized form of the Login message
```

Comment:
Because the signature is calculated against the serialized login. If we left Login typed here, we'd have to serialize once on the client to sign and then serialize another time on the server to check the signature, which requires both to produce byte perfect equal outputs (not a guarantee) and is slightly inefficient since we're serializing the same stuff twice.

Reply:
I see.  Thanks!